### PR TITLE
ApiListener patternmatching improvements

### PR DESCRIPTION
**Pull request by Yale96**

### CODE REVIEW

File: core/src/main/java/org/frankframework/http/rest/ApiServiceDispatcher.java [hunk 1]
```diff
@@ -15,44 +15,26 @@
...
-import jakarta.json.JsonObjectBuilder;
-import jakarta.json.JsonValue;
 import org.frankframework.align.XmlTypeToJsonSchemaConverter;
-import org.frankframework.core.IAdapter;
-import org.frankframework.core.IPipe;
-import org.frankframework.core.ListenerException;
-import org.frankframework.core.PipeLine;
-import org.frankframework.core.PipeLineExit;
-
+import org.frankframework.core.*;
```
Comment (nielsm5) [inline core/src/main/java/org/frankframework/http/rest/ApiServiceDispatcher.java:23]: Het is beter om geen wildcard imports te gebruiken. (Zie https://errorprone.info/bugpattern/WildcardImport)

File: core/src/main/java/org/frankframework/http/rest/ApiServiceDispatcher.java [hunk 2]
```diff
@@ -90,30 +72,46 @@ private List<ApiDispatchConfig>  findMatchingConfigsForUri(String uri, boolean e
...
 				if (exactMatch) {
 					return results;
 				}
 			}
+
+			if (uri.contains("**")) {
+				ApiDispatchConfig result = new ApiDispatchConfig(matchUri(uriPattern, uri));
+				results.add(result);
+				return results;
+			}
```
Comment (nielsm5) [inline core/src/main/java/org/frankframework/http/rest/ApiServiceDispatcher.java:102]: Je moet nooit zelf een nieuw patroon registreren, dat is de taak van de `registerServiceClient` methode.

File: core/src/test/java/org/frankframework/http/rest/ApiServiceDispatcherTest.java [hunk 3]
```diff
@@ -113,6 +114,22 @@ void testMultipleMethodsSameEndpointSameListener() throws Exception {
...
 	}
 
+	@ParameterizedTest
+	@CsvSource({
+			"/customers/**, /customers/123/addresses/345",
+			"/employees/**, /employees/123/departments/456/seats/52",
+			"/customers/**, /customers/123/addresses/345"
+	})
+	void testFindMatchDoubleAsterisk(String uriPattern, String expectedUri) throws ListenerException {
+		ApiListener listener = createServiceClient(ApiListenerServletTest.Methods.GET, expectedUri);
```
Comment (nielsm5) [inline core/src/test/java/org/frankframework/http/rest/ApiServiceDispatcherTest.java:124]: De ApiListeren zou moeten luisteren naar een url met een patroon, volgens mij test je het nu verkeerd om :P

**Review by nielsm5 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/frankframework/http/rest/ApiServiceDispatcher.java (hunk 1 above)
Comment (Yale96) [reply core/src/main/java/org/frankframework/http/rest/ApiServiceDispatcher.java:23]: Dit is aangepast, was een IntelliJ setting

File: core/src/test/java/org/frankframework/http/rest/ApiServiceDispatcherTest.java [hunk 4]
```diff
@@ -113,20 +116,90 @@ void testMultipleMethodsSameEndpointSameListener() throws Exception {
...
 		dispatcher.unregisterServiceClient(createServiceClient(ApiListenerServletTest.Methods.POST, uri));
-		ApiDispatchConfig config2 = dispatcher.findConfigForUri("/"+uri);
+		ApiDispatchConfig config2 = dispatcher.findExactMatchingConfigForUri("/"+uri);
 		assertNotNull(config2);
 		assertEquals("[GET]", config2.getMethods().toString());
 
 		//Test what happens after we remove both ServiceClient in the same DispatchConfig
 		dispatcher.unregisterServiceClient(createServiceClient(ApiListenerServletTest.Methods.GET, uri));
-		ApiDispatchConfig config3 = dispatcher.findConfigForUri("/"+uri);
+		ApiDispatchConfig config3 = dispatcher.findExactMatchingConfigForUri("/"+uri);
```
Comment (jkosternl) [inline core/src/test/java/org/frankframework/http/rest/ApiServiceDispatcherTest.java:202]: Super minor question: can you code-format these changed lines svp?

File: core/src/main/java/org/frankframework/http/rest/ApiServiceDispatcher.java [hunk 5]
```diff
@@ -74,54 +75,200 @@ public static synchronized ApiServiceDispatcher getInstance() {
...
+	 *     </lu>
+	 * </p>
+	 * @param uriPattern A pattern of a URI containing wildcards
+	 * @return Numerical score calculated for the URI based on the rules above.
+	 */
+	public static int scoreUriPattern(@Nonnull String uriPattern) {
+		// Scoring rules:
+		// - The more slashes the longer the match the more specific
+		// - The more wildcards in the pattern the less specific
+		// - "Match-all" patterns ending with /** are penalized with a -10 starting score
```
[the remark this answers is not in the document]
Comment (tnleeuw) [reply core/src/main/java/org/frankframework/http/rest/ApiServiceDispatcher.java:203]: You're right, I had these 3 lines before I had the JavaDoc. Forgot to clean it up after adding the JavaDoc.

File: pom.xml [hunk 6]
```diff
@@ -698,6 +698,9 @@
 						<release>${maven.compiler.target}</release>
 						<source>${maven.compiler.source}</source>
 						<target>${maven.compiler.target}</target>
+						<compilerArgs>
+							<arg>-parameters</arg>
```
[the remark this answers is not in the document]
Comment (tnleeuw) [reply pom.xml:702]: With this parameter, the bytecode produced by `javac` will include parameter names so you can see actual parameter names in reflection and in some exceptions instead of just `arg0`, `arg1`, etc.
I added that because some JUnit errors gave me just an abstract `arg3` that couldn't be matched.

Having actual names also sometimes helps Spring, for argument-matching for some parts of the framework, such as matching REST controller parameters to method arguments without specifying the name in each `@RequestParm` annotation etc.

**Comment by tnleeuw** [comment]
> I didn't expect that the change was so big or involved a new matching algorithm. I think it looks fine. Just a few minor remarks.

I didn't expect that either! But gradually that's where I ended up, having to take into account various edge-cases as you can find them in the unit tests.

File: pom.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (jkosternl) [reply pom.xml:702]: Ah clear, good idea to add this. Thanks for the explanation. Should we add a comment that `-parameters` has that effect? (or later in a different PR)

File: core/src/main/java/org/frankframework/http/rest/ApiListener.java [hunk 7]
```diff
@@ -414,14 +441,16 @@ public void setPrincipalNameClaim(String principalNameClaim) {
...
-		builder.append(" consumes["+getConsumes()+"]");
-		builder.append(" messageIdHeader["+getMessageIdHeader()+"]");
-		builder.append(" updateEtag["+isUpdateEtag()+"]");
-		return builder.toString();
+		return getClass().getSimpleName() + "@" + Integer.toHexString(hashCode()) +
+				" uriPattern[" + getUriPattern() + "]" +
+				" produces[" + getProduces() + "]" +
+				" consumes[" + getConsumes() + "]" +
+				" messageIdHeader[" + getMessageIdHeader() + "]" +
+				" updateEtag[" + isUpdateEtag() + "]";
```
Comment (nielsm5) [inline core/src/main/java/org/frankframework/http/rest/ApiListener.java:449]: Perhaps that it's me but, when do you prefer to use a StringBuilder vs 'normal' string concatenation?

File: core/src/main/java/org/frankframework/http/rest/ApiServiceDispatcher.java [hunk 8]
```diff
@@ -74,54 +75,196 @@ public static synchronized ApiServiceDispatcher getInstance() {
...
+	 *         <li>ApiListener2: GET on uri /user/{userId}/department/{departmentId}</li>
+	 *         <li>ApiListener3: POST on uri /user/{userId}/avatar</li>
+	 *     </lu>
+	 *     Then:
+	 *     <lu>
+	 *         <li>A request {@code GET /user/usr123/department/dept456} would return the {@link ApiDispatchConfig} for /user/{userId}/department/{departmentId} containing ApiListener2</li>
+	 *         <li>A request {@code OPTIONS /user/usr123/department/dept456} would return the {@link ApiDispatchConfig} for /user/{userId}/department/{departmentId} containing ApiListener2</li>
+	 *         <li>A request {@code GET /user/usr123/avatar} would return the {@link ApiDispatchConfig} for /user/** containing ApiListener1</li>
+	 *         <li>A request {@code POST /user/usr123/avatar} would return the {@link ApiDispatchConfig} for /user/{userId}/avatar containing ApiListener3</li>
+	 *         <li>A request {@code PUT /user/usr123/avatar} would return the {@link ApiDispatchConfig} for /user/{userId}/avatar containing ApiListener3</li>
```
Comment (nielsm5) [inline core/src/main/java/org/frankframework/http/rest/ApiServiceDispatcher.java:138]: But put has not been configured?

File: core/src/test/java/org/frankframework/http/rest/ApiListenerTest.java [hunk 9]
```diff
@@ -271,6 +273,29 @@ public void testValidatingMissingUriPattern() {
 		assertThrows(ConfigurationException.class, listener::configure, "uriPattern cannot be empty");
 	}
 
+	@ParameterizedTest
+	@CsvSource({
```
Comment (nielsm5) [inline core/src/test/java/org/frankframework/http/rest/ApiListenerTest.java:277]: 💯 expected outcome!

File: core/src/main/java/org/frankframework/http/rest/ApiListener.java [hunk 10]
```diff
@@ -198,14 +212,15 @@ private boolean hasMethod(HttpMethod method) {
...
 	 * @return null if no pattern is found
 	 */
 	public String getCleanPattern() {
 		String pattern = getUriPattern();
-		if(StringUtils.isEmpty(pattern))
+		if (StringUtils.isEmpty(pattern))
 			return null;
 
-		return pattern.replaceAll("\\{.*?}", "*");
+		return pattern.replaceAll("\\{[^}]*+}", "*");
```
[the remark this answers is not in the document]
Comment (tnleeuw) [reply core/src/main/java/org/frankframework/http/rest/ApiListener.java:223]: `[^}]1 is een character-class die alles matcht, behalve `}`. Deze is mij aangeraden door Sonar & IntelliJ als efficienter dan de oorspronkelijke regex, maar doet hetzelfde.

File: core/src/main/java/org/frankframework/http/rest/ApiListener.java [hunk 11]
```diff
@@ -256,20 +273,23 @@ public List<HttpMethod> getAllMethods() {
...
+	 *
 	 * @ff.mandatory
 	 */
 	public void setUriPattern(String uriPattern) {
-		if(!uriPattern.startsWith("/"))
-			uriPattern = "/" + uriPattern;
-
-		if(uriPattern.endsWith("/"))
-			uriPattern = uriPattern.substring(0, uriPattern.length()-1);
+		if (StringUtils.isNotBlank(uriPattern)) {
```
[the remark this answers is not in the document]
Comment (tnleeuw) [reply core/src/main/java/org/frankframework/http/rest/ApiListener.java:280]: Blank or not will be validated later, in `configure()`.
At this point, what this check prevents is that a blank string `""` will first be modified from `""` --> `"/"` and then from `"/"` again back to `""`.
This is what the old code would do on an empty `""` configuration string, before validating it's not empty.

File: core/src/main/java/org/frankframework/http/rest/ApiServiceDispatcher.java (hunk 8 above)
Comment (tnleeuw) [reply core/src/main/java/org/frankframework/http/rest/ApiServiceDispatcher.java:138]: And that is the point. I'll add more explicit explanation that this is for error-reporting purposes.

File: core/src/main/java/org/frankframework/http/rest/ApiListener.java (hunk 7 above)
Comment (tnleeuw) [reply core/src/main/java/org/frankframework/http/rest/ApiListener.java:449]: `StringBuilder` when there are conditionals or loops but string concatenation when it is all a single statement.

File: core/src/test/java/org/frankframework/http/rest/ApiListenerTest.java (hunk 9 above)
Comment (tnleeuw) [reply core/src/test/java/org/frankframework/http/rest/ApiListenerTest.java:277]: I'm not sure what you mean?

**Review by jkosternl — APPROVED** [review]

File: core/src/main/java/org/frankframework/http/rest/ApiListener.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (nielsm5) [reply core/src/main/java/org/frankframework/http/rest/ApiListener.java:280]: Ah, perhaps we could validate in configure if this has been set, I think it's best to not allow people to use the `/` (aka root) directly.

File: core/src/test/java/org/frankframework/http/rest/ApiListenerTest.java (hunk 9 above)
Comment (nielsm5) [reply core/src/test/java/org/frankframework/http/rest/ApiListenerTest.java:277]: It's a good test!

**Review by nielsm5 — APPROVED** [review]

**Review by nielsm5 — APPROVED** [review]
